rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,92 @@
+---
+feature: nix-language-version
+start-date: 2022-12-12
+author: @fricklerhandwerk
+co-authors: @thufschmitt @Ericson2314
+shepherd-team: 
+shepherd-leader: 
+related-issues: https://github.com/NixOS/nix/issues/7255
+---
+
+# Summary
+[summary]: #summary
+
+Introduce a convention to determine which version of the Nix language grammar to use for parsing a Nix file.
+
+# Motivation
+[motivation]: #motivation
+
+Currently it's impossible to introduce backwards-incompatible changes to the Nix language without breaking existing setups.
+This proposal is first step towards overcoming that limitation.
+
+# Detailed design
+[design]: #detailed-design
+
+Introduce a magic comment in the first line of a Nix file:
+
+    #? Nix <version>
+
+where `<version>` is a released version of Nix the given file is intended to work with.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+```nix
+#? Nix 2.12
+"nothing"
+```
+
+# Arguments
+[advantages]: #advantages
+
+* (+) Encourages developing Nix to deal with changes to the Nix language in a principled manner.
```

Comment:
How does this help with disciplined changes? I can only see that it makes experiments in production and easier as you can retire them with the next version. 

This sounds like a feature in practice but nix will need to keep all the older features. The code in its current form will suffer even more if it isn't just a weird language but also one with a few quirks being enabled and disabled with every other version.

Reply:
One way I can see around this is to require that for any version change, there exists a purely syntactical transformation that upgrades some code to the next version. We would want this anyways in order to build automatic upgrade tooling. But requiring it means that the core nix code can only worry about the latest version and the backwards compatibility can be stuffed into a corner where it is not in the way most of the time.

Reply:
You're right. Updated the arguments accordingly.

Reply:
> One way I can see around this is to require that for any version change, there exists a purely syntactical transformation that upgrades some code to the next version. We would want this anyways in order to build automatic upgrade tooling. But requiring it means that the core nix code can only worry about the latest version and the backwards compatibility can be stuffed into a corner where it is not in the way most of the time.

This is a) doubtful (there won't only be syntactic changes) and b) not feasible, as nixpkgs *needs* to be evaluatable by multiple Nix versions.

Reply:
a) is a question of scope. It would be perfectly fine to decide on making that restriction for the language version feature.

b) I don't see how this is specific to my proposal compared to the entire RFC

-------

Anyways. Let's continue the discussion of my idea at https://github.com/NixOS/rfcs/pull/137#discussion_r1052409918